File: src/lib/api/assetSearch.ts
/** API helper for Asset Search */

export interface AssetSearchRequest {
  query: string;
}

export interface AssetSearchResponse {
  name: string;
  ticker?: string;
  isin?: string;
  country?: string;
  region?: string;
  sub_region?: string;
  sector?: string;
  asset_class?: string;
  description?: string;
  confidence: number;
}

const API_BASE = process.env.NEXT_PUBLIC_API_BASE_URL?.trim() || "http://localhost:8000";

export async function searchAsset(query: string): Promise<AssetSearchResponse> {
  try {
    const response = await fetch(`${API_BASE}/asset-search/search`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ query }),
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(`Asset search failed: ${response.status} ${errorText}`);
    }

    const data = await response.json();
    return data as AssetSearchResponse;
  } catch (error) {
    if (error instanceof Error) {
      throw error;
    }
    throw new Error("Failed to search for asset");
  }
}

File: backend/routes/asset_search.py
"""FastAPI routes for asset search."""
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Optional

from ..asset_search import search_asset, AssetInfo


router = APIRouter(prefix="/asset-search", tags=["asset-search"])


class AssetSearchRequest(BaseModel):
    """Asset search request."""
    query: str


class AssetSearchResponse(BaseModel):
    """Asset search response."""
    name: str
    ticker: Optional[str]
    isin: Optional[str]
    country: Optional[str]
    region: Optional[str]
    sub_region: Optional[str]
    sector: Optional[str]
    asset_class: Optional[str]
    description: Optional[str]
    confidence: float


@router.post("/search")
def search_for_asset(request: AssetSearchRequest) -> AssetSearchResponse:
    """
    Search for an asset using web search + Claude Haiku extraction.

    Args:
        request: AssetSearchRequest with query string

    Returns:
        AssetSearchResponse with extracted asset information
    """
    if not request.query or len(request.query.strip()) < 2:
        raise HTTPException(status_code=400, detail="Query must be at least 2 characters")

    # Search for asset
    asset_info = search_asset(request.query.strip())

    if not asset_info:
        raise HTTPException(status_code=404, detail="Asset not found")

    return AssetSearchResponse(
        name=asset_info.name,
        ticker=asset_info.ticker,
        isin=asset_info.isin,
        country=asset_info.country,
        region=asset_info.region,
        sub_region=asset_info.sub_region,
        sector=asset_info.sector,
        asset_class=asset_info.asset_class,
        description=asset_info.description,
        confidence=asset_info.confidence,
    )
